game_input/src/lib.rs

```diff
@@ -1,22 +1,21 @@
 extern crate game_window;
 extern crate game_logic;
 
-use std::iter::IteratorExt;
 use game_logic::LogicState;
 use game_window::{Action, Key, WindowEvent};
 
 pub type InputHandler = fn(LogicState, WindowEvent) -> LogicState;
 
-pub fn handle_events(logic_state: LogicState, events: Vec<WindowEvent>) -> LogicState {
-  events.iter().fold(logic_state, check_keys)
-}
-
-fn check_keys(logic_state: LogicState, event: &WindowEvent) -> LogicState {
-  match *event {
-    WindowEvent::Key(Key::Escape, _, Action::Press, _) => logic_state.as_closed(),
-    WindowEvent::Key(Key::A, _, Action::Press, _) => logic_state.inc_edges(),
-    WindowEvent::Key(Key::D, _, Action::Press, _) => logic_state.dec_edges(),
-    _ => logic_state
+pub fn handle_events(logic_state: &mut LogicState, events: Vec<WindowEvent>) {
+  for event in events.iter() {
+    match *event {
+      WindowEvent::Key(Key::Escape, _, Action::Press, _) => logic_state.closed(),
+      WindowEvent::Key(Key::A, _, Action::Press, _) => logic_state.inc_edges(),
+      WindowEvent::Key(Key::D, _, Action::Press, _) => logic_state.dec_edges(),
+      WindowEvent::Key(Key::A, _, Action::Repeat, _) => logic_state.inc_edges(),
+      WindowEvent::Key(Key::D, _, Action::Repeat, _) => logic_state.dec_edges(),
+      _ => ()
+    }
```

Comment:
You can still fold over this as you were doing, by taking an &LogicState instead of an &mut LogicState, then clone()ing it and folding like normal, returning a new LogicState, if you really wanted to stay away from side effects.
Also rust match arms can use | to concatenate identical result arms:
```rust
use game_window::WindowEvent::*;
use game_window::Key::*;
use game_window::Action::*;
match *event {
  Key(Escape, _, Press, _) => logic_state.closed(),
  Key(A, _, Press, _) | Key(A, _, Repeat, _) => logic_state.inc_edges(),
  Key(D, _, Press, _) | Key(D, _, Repeat, _) => logic_state.dec_edges(),
  _ => ()
}
```